src/S3/Transfer.php

```diff
@@ -109,19 +109,32 @@ public function __construct(
     }
 
     /**
-     * Creates an iterator that recursively yields all filenames in a directory
+     * Returns a recursive directory iterator that yields filenames only and
+     * is not broken like PHP's built-in DirectoryIterator (which will read
+     * the first file from a stream wrapper, then rewind, then read it again).
      *
      * @param string $path Path on disk to traverse
      *
      * @return \Iterator Returns an iterator that yields absolute filenames.
      */
-    public static function fileIterator($path)
+    public static function recursiveDirIterator($path)
```

Comment:
Here I've created my own implementation of RecursiveDirectoryIterator that doesn't have weird rewind semantics like PHP's built-in RecursiveDirectoryIterator. The built-in iterator will immediately read the first file using dir_readdir, then it will be rewound and call dir_rewinddir when you begin iterating. This caused the first element of our generators to be skipped.

Here I've just reimplemented RecursiveDirectoryIterator using a generator and a queue of values to yield. This works on any type of stream wrapper and does not call unnecessary rewinds.